### PR TITLE
[native]Improve spill directory failure handling

### PR DESCRIPTION
**Pull request by xiaoxmeng**
There are three problems found in Meta internal arbitration and spilling test about
task spilling directory setup:
(1) velox task creation failure handle path doesn't clear velox task before allowing
create the next velox task which can lead to the second velox task creation (retry)
fail with unrelated errors;
(2) set spilling path before the spill path creation succeed. The better is to create the
spilling directory and then set it task to avoid another exception on task destruction
which removes a non-exist spill directory/

This PR fixes this issue with unit test. Note the spill directory setup could fail because
of transient remote storage error as we found in Meta internal testing.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/TaskManager.cpp [hunk 1]
```diff
@@ -42,30 +42,37 @@ namespace {
...
+        velox::filesystems::getFileSystem(taskSpillDirPath, nullptr);
+    VELOX_CHECK_NOT_NULL(fileSystem, "File System is null!");
+    fileSystem->mkdir(taskSpillDirPath);
+  } catch (...) {
+    LOG(WARNING) << "Failed to create spill directory for velox task "
+                 << veloxTask.taskId() << ". Error: "
+                 << folly::exceptionStr(std::current_exception());
+    return;
+  }
+  veloxTask.setSpillDirectory(taskSpillDirPath);
```
Comment (tanjialiang) [inline presto-native-execution/presto_cpp/main/TaskManager.cpp:75]: Thanks, curious what will happen when path creation fails and we still set the path.

**Review by tanjialiang — APPROVED** [review]

File: presto-native-execution/presto_cpp/main/TaskManager.cpp (hunk 1 above)
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/TaskManager.cpp:75]: It will get another exception on rmdir and might distract the debug but the query fails so it won't get executed.

File: presto-native-execution/presto_cpp/main/TaskManager.cpp [hunk 2]
```diff
@@ -42,30 +42,37 @@ namespace {
...
+    // Create folder for the task spilling.
+    auto fileSystem =
+        velox::filesystems::getFileSystem(taskSpillDirPath, nullptr);
+    VELOX_CHECK_NOT_NULL(fileSystem, "File System is null!");
+    fileSystem->mkdir(taskSpillDirPath);
+  } catch (...) {
+    LOG(WARNING) << "Failed to create spill directory for velox task "
+                 << veloxTask.taskId() << ". Error: "
+                 << folly::exceptionStr(std::current_exception());
+    return;
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/TaskManager.cpp:73]: Why is it OK to ignore this error? If we fail to setup spill directory and the query runs out of memory, will we get some cryptic error after wasting a lot of compute?

File: presto-native-execution/presto_cpp/main/TaskManager.cpp [hunk 3]
```diff
@@ -424,7 +431,7 @@ std::unique_ptr<TaskInfo> TaskManager::createOrUpdateTask(
     const protocol::OutputBuffers& outputBuffers,
     std::shared_ptr<velox::core::QueryCtx> queryCtx,
     long startProcessCpuTime) {
-  std::shared_ptr<exec::Task> execTask;
+  std::shared_ptr<exec::Task> veloxTask;
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/TaskManager.cpp:434]: Nice rename, but it would be better to do this in a separate PR. Mixing trivial refactoring with meaningful changes makes it hard to review the code.

File: presto-native-execution/presto_cpp/main/TaskManager.cpp [hunk 4]
```diff
@@ -436,21 +443,24 @@ std::unique_ptr<TaskInfo> TaskManager::createOrUpdateTask(
         return std::make_unique<TaskInfo>(prestoTask->updateInfoLocked());
       }
 
-      execTask = exec::Task::create(
-          taskId, planFragment, prestoTask->id.id(), std::move(queryCtx));
-      auto baseSpillDir = *(baseSpillDir_.rlock());
-      maybeSetupTaskSpillDirectory(planFragment, *execTask, baseSpillDir);
+      {
+        auto newVeloxTask = exec::Task::create(
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/TaskManager.cpp:447]: Why do we need this variable?

File: presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp [hunk 5]
```diff
@@ -1132,6 +1132,67 @@ TEST_F(TaskManagerTest, checkBatchSplits) {
...
+      const std::string badSpillPath = "/nonexist/presto_native/path";
+      taskManager_->setBaseSpillDirectory(badSpillPath);
+    }
+
+    std::map<std::string, std::string> queryConfigs;
+    queryConfigs.emplace(core::QueryConfig::kTestingSpillPct, "100");
+    queryConfigs.emplace(core::QueryConfig::kSpillEnabled, "true");
+    queryConfigs.emplace(core::QueryConfig::kAggregationSpillEnabled, "true");
+
+    auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp:1167]: planNodeIdGenerator is not needed for plans without joins

File: presto-native-execution/presto_cpp/main/TaskManager.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/TaskManager.cpp:73]: We currently create spilling directory at the startup of each query so the query might not spill. So instead of failing query, we can proceed the query without spilling enabled. We could later add a thread-safe method at query level to create a spilling directory when needs if this happens too often in practice.

Comment (mbasmanova) [reply presto-native-execution/presto_cpp/main/TaskManager.cpp:73]: I'm not sure I understand. What would cause a failure to setup spill directory? If this is cause by bad config, where base spill directory doesn't exist than it might be better to fail early, no?

File: presto-native-execution/presto_cpp/main/TaskManager.cpp (hunk 4 above)
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/TaskManager.cpp:447]: This is to destroy the velox task under presto task lock so as to ensure the second presto task update call won't see the leftover state from the previous failed velox task such as the task memory pool under query root memory pool. Here is the event sequence:
T1: task manager receive request from coordinator to create velox task;
T2: task manager holds the presto task lock to create a new velox task;
T3: the spilling directory setup fails and throw, but the new velox task hans't destroyed yet;
T4: task manager releases the presto task lock and second create velox task request comes;
T5: task manager creates a new velox task for the same presto task but the memory pool of the previously created velox task still stays in the query root memory pool;
T6: the second velox task creation fails because of the duplicate task memory pool name.

File: presto-native-execution/presto_cpp/main/TaskManager.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/TaskManager.cpp:73]: Not bad config. From the failure that we have been seen in Meta internal testing, it is caused by some remote storage error (some internal timeout at storage backend). The spilling directory setup happens on the query start but not when we trigger spills.

File: presto-native-execution/presto_cpp/main/TaskManager.cpp (hunk 4 above)
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/TaskManager.cpp]: I removed it as maybeSetupDirectory now tolerates the failure.

File: presto-native-execution/presto_cpp/main/TaskManager.cpp (hunk 3 above)
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/TaskManager.cpp]: Reverted.

File: presto-native-execution/presto_cpp/main/TaskManager.cpp [hunk 6]
```diff
@@ -52,20 +52,27 @@ static void maybeSetupTaskSpillDirectory(
...
       nodeConfig->nodeId(),
       execTask.queryCtx()->queryId(),
       execTask.taskId(),
       includeNodeInSpillPath);
+
+  try {
+    // Create folder for the task spilling.
+    auto fileSystem =
+        velox::filesystems::getFileSystem(taskSpillDirPath, nullptr);
+    VELOX_CHECK_NOT_NULL(fileSystem, "File System is null!");
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/TaskManager.cpp:67]: Do we want to catch and ignore this failure as well?

File: presto-native-execution/presto_cpp/main/TaskManager.cpp [hunk 7]
```diff
@@ -52,20 +52,27 @@ static void maybeSetupTaskSpillDirectory(
...
+    fileSystem->mkdir(taskSpillDirPath);
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/TaskManager.cpp:68]: Maybe add a comment here to explain that this might be a transient failure and we want to process with task creation and execution hoping spilling won't be needed.

However, wouldn't it be confusing if queries fail with OOM errors while cluster is configured with spilling enabled?

File: presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp [hunk 8]
```diff
@@ -1132,6 +1132,66 @@ TEST_F(TaskManagerTest, checkBatchSplits) {
   ASSERT_EQ(resultOrFailure.status, nullptr);
 }
 
+TEST_F(TaskManagerTest, spillDirectoryNonExist) {
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp:1135]: nit: NotExist -> Missing

File: presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp [hunk 9]
```diff
@@ -1132,6 +1132,66 @@ TEST_F(TaskManagerTest, checkBatchSplits) {
...
+  // Cleanup old tasks between test iterations.
+  taskManager_->setOldTaskCleanUpMs(0);
+  for (bool spillDirExist : {false, true}) {
+    SCOPED_TRACE(fmt::format("spillDirExist: {}", spillDirExist));
+    auto spillDir = setupSpillPath();
+
+    const int numBatchesPerFile = 5;
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp:1142]: Do we need 5 batches of 1000 rows each and 5 files? Can we make 1 batch with 10 rows and use Values node instead of TableScan? This will make this test run faster.

File: presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp [hunk 10]
```diff
@@ -1132,6 +1132,66 @@ TEST_F(TaskManagerTest, checkBatchSplits) {
...
+      taskManager_->setBaseSpillDirectory(
+          fmt::format("{}/presto_native/path", spillDir->path));
+    } else {
+      // Set bad spill path.
+      const std::string badSpillPath = "/nonexist/presto_native/path";
+      taskManager_->setBaseSpillDirectory(badSpillPath);
+    }
+
+    std::map<std::string, std::string> queryConfigs;
+    queryConfigs.emplace(core::QueryConfig::kTestingSpillPct, "100");
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp:1163]: Would it make sense to extract a helper function to "enable spilling"?

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng Thank you for explaining and making the PR easier to review. Looks good % a couple of questions.

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng Would you update commit message to provide more details about the changes?

File: presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp (hunk 9 above)
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp:1142]: Not required

File: presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp (hunk 10 above)
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp:1163]: Sure.

File: presto-native-execution/presto_cpp/main/TaskManager.cpp [hunk 11]
```diff
@@ -436,20 +435,22 @@ std::unique_ptr<TaskInfo> TaskManager::createOrUpdateTask(
         return std::make_unique<TaskInfo>(prestoTask->updateInfoLocked());
       }
 
-      execTask = exec::Task::create(
+      auto newExecTask = exec::Task::create(
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/TaskManager.cpp:438]: nit: maybe add a comment to explain the use a separate variable

File: presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp [hunk 12]
```diff
@@ -1132,6 +1138,71 @@ TEST_F(TaskManagerTest, checkBatchSplits) {
   ASSERT_EQ(resultOrFailure.status, nullptr);
 }
 
+TEST_F(TaskManagerTest, buildSpillDirectryFailure) {
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp:1141]: typo: Directry -> Directory

File: presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp [hunk 13]
```diff
@@ -1132,6 +1138,71 @@ TEST_F(TaskManagerTest, checkBatchSplits) {
...
+  // Cleanup old tasks between test iterations.
+  taskManager_->setOldTaskCleanUpMs(0);
+  for (bool buildSpillDirectoryFailure : {false}) {
+    SCOPED_TRACE(fmt::format(
+        "buildSpillDirectoryFailure: {}", buildSpillDirectoryFailure));
+    auto spillDir = setupSpillPath();
+
+    auto filePaths = makeFilePaths(1);
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp:1149]: Why do we need to write data to files? Can we not use values(batches) node in the plan?

File: presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp [hunk 14]
```diff
@@ -1132,6 +1138,71 @@ TEST_F(TaskManagerTest, checkBatchSplits) {
...
+    updateRequest.session.systemProperties = queryConfigs;
+    // Create task will fail if the spilling directory setup fails.
+    if (buildSpillDirectoryFailure) {
+      VELOX_ASSERT_THROW(
+          createOrUpdateTask(taskId, updateRequest, planFragment), "Mkdir");
+    } else {
+      createOrUpdateTask(taskId, updateRequest, planFragment);
+    }
+    {
+      auto taskMap = taskManager_->tasks();
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/tests/TaskManagerTest.cpp:1182]: Maybe move this block into 'else' branch above.

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Please, update PR description to match the code. Consider explaining why directory creation may fail as it is non-intuitive. Please, update commit message to add details.
